build-data-capturing-gradle-samples/capture-thermal-throttling/gradle-thermal-throttling.gradle

```diff
@@ -0,0 +1,139 @@
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.os.OperatingSystem
+
+import java.nio.charset.Charset
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+
+/**
+ * This Gradle script captures the thermal throttling level and adds it as a tag.
+ * Some parameters can be tweaked according to your build:
+ * - SAMPLING_INTERVAL_IN_SECONDS, frequency at which the capture command is run
+ * - THROTLLING_LEVEL, list of throttling levels and value ranges (to be compared with captured throttling average value)
+ *
+ * WARNINGS:
+ * - This is supported on MacOS only.
+ *
+ * This should be applied to the root project:
+ * <code> apply from: file('gradle-thermal-throttling.gradle') </code>
+ */
+
+def buildScanApi = project.extensions.findByName('buildScan')
+if (!buildScanApi) {
+  return
+}
+
+// Register Thermal throttling service
+def throttlingServiceProvider = gradle.sharedServices.registerIfAbsent("thermalThrottling", ThermalThrottlingService.class) {}
+
+// Start Thermal throttling service
+throttlingServiceProvider.get()
+
+// Process results on build completion
+buildScan.buildFinished { def result ->
+  throttlingServiceProvider.get().processResults(buildScanApi)
+}
+
+// Thermal Throttling service implementation
+abstract class ThermalThrottlingService implements BuildService<BuildServiceParameters.None>, AutoCloseable {
+
+  /**
+   * Sampling interval can be adjusted according to total build time.
+   */
+  def static final SAMPLING_INTERVAL_IN_SECONDS = 5
```

Comment:
are the `final` here needed? our other scripts don't use `final` or `static`? I think it would be good to keep them in the same consistent simple style.

Reply:
It's an habit to ensure [safe initialization](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5)
I think we can remove the keywords to get an alignment, the concurrency shouldn't be much of an issue for this managed object.

Reply:
Yes, I think we should remove these to simplify these scripts and keep the style consistent.

Reply:
> Yes, I think we should remove these to simplify these scripts and keep the style consistent.

Agreed, this is done in local.  
Unrelated, but I discovered that when CC is restored the service is not triggered, I'm investigating this before moving forward

Reply:
I eventually encapsulated the service start in a `buildScan.background` block to make sure it would be triggered at the beginning of the build in any situation (without CC, with CC when storing entry, with CC when restoring entry).

Having the start service instruction at the root level of the script doesn't work when restoring a CC entry, in this case the service is only instantiated when the service is required in the `buildScan.buildFinished` block.